exchanges/stream/websocket.go

```diff
@@ -576,183 +555,116 @@ func (w *Websocket) trafficMonitor() {
 	w.Wg.Add(1)
 
 	go func() {
-		var trafficTimer = time.NewTimer(w.trafficTimeout)
-		pause := make(chan struct{})
+		t := time.NewTimer(w.trafficTimeout)
```

Comment:
suggestion: don't need a timer can pivot around a time variable, it could potential t.Stop calls and drain, but either ways yours works 🤷 so can ignore.

Reply:
I don't see a clear path to that suggestion.
If t becomes time.Time then line 579 needs to use a channel somehow.
If 579 becomes time.After() it'd need to subtract Now() from the time declared at 558, which feels messy.
I scanned through time pkg for any other channel methods than time.After() and couldn't see any.

Happy to change it if there's an improvement, but I can't see it yet.

Note: I tried variations with timer and ticker several times. This was the least convoluted solution I could spot.

Reply:
This is what I meant:
```diff
diff --git a/exchanges/stream/websocket.go b/exchanges/stream/websocket.go
index 7d1d402d1..f77ad308a 100644
--- a/exchanges/stream/websocket.go
+++ b/exchanges/stream/websocket.go
@@ -555,27 +555,26 @@ func (w *Websocket) trafficMonitor() {
        w.Wg.Add(1)

        go func() {
-               t := time.NewTimer(w.trafficTimeout)
+               defer func() {
+                       w.setTrafficMonitorRunning(false)
+                       w.Wg.Done()
+                       if !w.IsConnected() {
+                               return
+                       }
+                       if err := w.Shutdown(); err != nil {
+                               log.Errorf(log.WebsocketMgr, "%v websocket: trafficMonitor shutdown err: %s", w.exchangeName, err)
+                       }
+               }()
+
+               next := time.Now().Add(w.trafficTimeout)
                for {
                        select {
                        case <-w.ShutdownC:
                                if w.verbose {
                                        log.Debugf(log.WebsocketMgr, "%v websocket: trafficMonitor shutdown message received", w.exchangeName)
                                }
-                               t.Stop()
-                               w.setTrafficMonitorRunning(false)
-                               w.Wg.Done()
                                return
                        case <-time.After(trafficCheckInterval):
-                               select {
-                               case <-w.TrafficAlert:
-                                       if !t.Stop() {
-                                               <-t.C
-                                       }
-                                       t.Reset(w.trafficTimeout)
-                               default:
-                               }
-                       case <-t.C:
                                checkAgain := w.IsConnecting()
                                select {
                                case <-w.TrafficAlert:
@@ -583,20 +582,15 @@ func (w *Websocket) trafficMonitor() {
                                default:
                                }
                                if checkAgain {
-                                       t.Reset(w.trafficTimeout)
+                                       next = time.Now().Add(w.trafficTimeout)
+                                       break
+                               }
+                               if time.Now().Before(next) {
                                        break
                                }
                                if w.verbose {
                                        log.Warnf(log.WebsocketMgr, "%v websocket: has not received a traffic alert in %v. Reconnecting", w.exchangeName, w.trafficTimeout)
                                }
-                               w.setTrafficMonitorRunning(false) // Cannot defer lest Connect is called after Shutdown but before deferred call
-                               w.Wg.Done()                       // Without this the w.Shutdown() call below will deadlock
-                               if w.IsConnected() {
-                                       err := w.Shutdown()
-                                       if err != nil {
-                                               log.Errorf(log.WebsocketMgr, "%v websocket: trafficMonitor shutdown err: %s", w.exchangeName, err)
-                                       }
-                               }
                                return
                        }
                }
diff --git a/exchanges/stream/websocket_test.go b/exchanges/stream/websocket_test.go
index b099adca1..0ce64edca 100644
--- a/exchanges/stream/websocket_test.go
+++ b/exchanges/stream/websocket_test.go
@@ -280,7 +280,7 @@ func TestTrafficMonitorShutdown(t *testing.T) {
        case <-time.After(1 * trafficCheckInterval):
        }

-       close(ws.ShutdownC)
+       require.NoError(t, ws.Shutdown())

        <-time.After(2 * trafficCheckInterval)
        assert.False(t, ws.IsTrafficMonitorRunning(), "traffic monitor should be shutdown")
```